### ISSUE TITLE
Implement model parallel decoding for top-k, top-p, and beam search

### ISSUE THREAD

**Issue by gcooper-isi** [gcooper-isi/Megatron-LM#18]
#13 includes an implementation of greedy decoding that operates in parallel on each device within a model parallel group, in contrast with the previous implementation, which first combines the logits from each device before performing the decoding. By not creating a combined logits tensor, you can save a significant amount of GPU memory. This task is to implement similar model parallel decoding for top-k, top-p, and beam search as well.
